src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Logging/UnconditionalSuppressMessageAttributeState.cs

```diff
@@ -0,0 +1,338 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection.Metadata;
+
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+using ILCompiler.Dataflow;
+
+using ILLink.Shared;
+
+#nullable enable
+
+namespace ILCompiler.Logging
+{
+    public class UnconditionalSuppressMessageAttributeState
+    {
+        internal const string ScopeProperty = "Scope";
+        internal const string TargetProperty = "Target";
+        internal const string MessageIdProperty = "MessageId";
+
+        internal const string UnconditionalSuppressMessageAttributeNamespace = "System.Diagnostics.CodeAnalysis";
+        internal const string UnconditionalSuppressMessageAttributeName = "UnconditionalSuppressMessageAttribute";
+
+        public class Suppression
+        {
+            public SuppressMessageInfo SuppressMessageInfo { get; }
+            public bool Used { get; set; }
+            public CustomAttributeValue<TypeDesc> OriginAttribute { get; }
+            public TypeSystemEntity Provider { get; }
+
+            public Suppression(SuppressMessageInfo suppressMessageInfo, CustomAttributeValue<TypeDesc> originAttribute, TypeSystemEntity provider)
+            {
+                SuppressMessageInfo = suppressMessageInfo;
+                OriginAttribute = originAttribute;
+                Provider = provider;
+            }
+        }
+
+        private readonly Dictionary<TypeSystemEntity, Dictionary<int, Suppression>> _suppressions;
+        private HashSet<IAssemblyDesc> InitializedAssemblies { get; }
+        private readonly CompilerGeneratedState? _compilerGeneratedState;
+        private readonly Logger _logger;
+
+        public UnconditionalSuppressMessageAttributeState(CompilerGeneratedState? compilerGeneratedState, Logger logger)
+        {
+            _suppressions = new Dictionary<TypeSystemEntity, Dictionary<int, Suppression>>();
```

Comment:
Do we have any thread safety concerns?

Reply:
Great point - sorry, still in the single-threaded illink mindset.